creusot/src/backend/ty.rs

```diff
@@ -514,12 +514,20 @@ pub(crate) fn destructor<'tcx>(
         ),
     };
 
-    let fail = Expr::Assert(Box::new(Exp::mk_false()), Box::new(Expr::Any));
+    let fail = Expr::BlackBox(Box::new(Expr::Assert(Box::new(Exp::mk_false()), Box::new(Expr::Any))));
+
+    let existential_params : Vec<_> = fields.iter().cloned().collect();
+    let negative_assertion = if existential_params.is_empty() {
+        cons_test.neq(Exp::var("input"))
+    } else {
+         Exp::Exists(existential_params, vec![], Box::new(cons_test.eq(Exp::var("input")))).not()
+    };
```

Comment:
Not Exists x, P  => Forall x, Not P

```suggestion
    let forall_params : Vec<_> = fields.iter().cloned().collect();
    let negative_assertion = if forall_params_params.is_empty() {
        cons_test.neq(Exp::var("input"))
    } else {
         Exp::Forall(forall_params, vec![], Box::new(cons_test.neq(Exp::var("input"))))
    };
```

Also, we could add `cons_test` as a trigger, because this is the obvious thing to do here. The SMT solvers should be able to do that by themselves, but who knows what Why3 does with quantifiers and polymorphism...

Reply:
yea talking with @paulpatault he suggested doing the exists -> forall transformation. my incoming update includes that.